Integration of circuit ansatzes with `qibo.models.encodings`

chmwzc opened the pull request:
Thanks again to @renatomello for raising #127

Have refactored: `hf_circuit` and `he_circuit`
Yet to investigate: UCC with `gates.RBS` (or `gates.GIVENS`) and `gates.GeneralizedRBS`

Some other things that I noticed when looking through the code in `qibo.models.encodings` that we can try to add in as well:
- Add keywords arguments for `qibo.Circuit`
- Error handling: `from qibo.config import raise_error`

In the longer term, if we do look into implementing excited state ansatzes, it might be cleaner to combine all the current (ground state) ansatzes into a single module? Just a thought for the moment

chmwzc committed:
Use comp_basis_encoder to build HF circuit

chmwzc committed:
Use `entangling_layer` function for HE ansatz

chmwzc committed:
Merge branch 'samp_var_v2' into encodings

chmwzc committed:
Minor changes and cleanup

chmwzc committed:
Fix pylint error

chmwzc committed:
Fix tests

renatomello commented on src/qibochem/ansatz/hardware_efficient.py:
```diff
@@ -3,36 +3,55 @@
 """
 
 from qibo import Circuit, gates
```
```suggestion
from typing import Optional

from qibo import Circuit, gates
```

renatomello commented on src/qibochem/ansatz/hardware_efficient.py:
```diff
@@ -3,36 +3,55 @@
+    :class:`qibo.models.encodings.entangling_layer`.
 
     Args:
         n_qubits (int): Number of qubits in the quantum circuit
         n_layers (int): Number of layers of rotation and entangling gates
```
```suggestion
        nqubits (int): Number of qubits in the quantum circuit.
        nlayers (int): Number of layers of rotation and entangling gates.
```

renatomello commented on src/qibochem/ansatz/hardware_efficient.py:
```diff
@@ -3,36 +3,55 @@
 
-    circuit = Circuit(n_qubits)
-
+    circuit = Circuit(n_qubits, **kwargs)
     for _ in range(n_layers):
```
```suggestion
    circuit = Circuit(nqubits, **kwargs)
    for _ in range(nlayers):
```

renatomello commented on src/qibochem/ansatz/hardware_efficient.py:
```diff
@@ -3,36 +3,55 @@
-        circuit.add(getattr(gates, coupling_gates)(qubit, qubit + 1) for qubit in range(n_qubits - 1))
-        circuit.add(getattr(gates, coupling_gates)(n_qubits - 1, 0))
-
+        circuit.add(
+            rgate(qubit, theta=0.0)  # pylint: disable=not-callable
+            for qubit in range(n_qubits)
+            for rgate in parameter_gates
+        )
+        # Entangling gates
+        circuit += entangling_layer(n_qubits, architecture, entangling_gate, closed_boundary, **kwargs)
```
```suggestion
        circuit.add(
            rgate(qubit, theta=0.0)  # pylint: disable=not-callable
            for qubit in range(nqubits)
            for rgate in parameter_gates
        )
        # Entangling gates
        circuit += entangling_layer(nqubits, architecture, entangling_gate, closed_boundary, **kwargs)
```

chmwzc committed:
Update src/qibochem/ansatz/hardware_efficient.py

Co-authored-by: Renato M S Farias <renato.msf@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

chmwzc committed:
Update src/qibochem/ansatz/hardware_efficient.py

Co-authored-by: Renato M S Farias <renato.msf@gmail.com>

chmwzc committed:
Update src/qibochem/ansatz/hardware_efficient.py

Co-authored-by: Renato M S Farias <renato.msf@gmail.com>

chmwzc committed:
Update src/qibochem/ansatz/hardware_efficient.py

Co-authored-by: Renato M S Farias <renato.msf@gmail.com>

chmwzc committed:
Update src/qibochem/ansatz/hardware_efficient.py

Co-authored-by: Renato M S Farias <renato.msf@gmail.com>

renatomello commented on src/qibochem/ansatz/hf_reference.py:
```diff
@@ -1,83 +1,81 @@
+
+
+# Helper functions for the Brayvi-Kitaev mapping
+def _bk_matrix_power2(n: int):
+    """Build the Brayvi-Kitaev matrix of dimension n (= power of 2) recursively
+
+    Args:
+        n: Size of BK matrix
+    """
+    assert n > 0, "Dimension of BK matrix must be at least 1"
+    # Base case
+    if n == 1:
+        return np.ones((1, 1), dtype=np.int8)
+    # Recursive definition
+    smaller_bk_matrix = _bk_matrix_power2(n - 1)
+    top_right = np.zeros((2 ** (n - 2), 2 ** (n - 2)), dtype=np.int8)
+    top_half = np.concatenate((smaller_bk_matrix, top_right), axis=1)
+
+    bottom_left = np.concatenate(
+        (np.zeros(((2 ** (n - 2)) - 1, 2 ** (n - 2)), dtype=np.int8), np.ones((1, 2 ** (n - 2)), dtype=np.int8)), axis=0
+    )
+    bottom_half = np.concatenate((bottom_left, smaller_bk_matrix), axis=1)
+    # Combine top and bottom half
+    return np.concatenate((top_half, bottom_half), axis=0)
```
```suggestion
def _bk_matrix_power2(dims: int):
    """Build the Bravyi-Kitaev matrix of dimension ``dims`` :math:`d = 2^{n}` recursively.

    Args:
        dims (int): Size of Bravyi-Kitaev matrix.

    Returns:
        ndarray: Bravyi-Kitaev matrix of size ``dims x dims``.
    """
    assert dims > 0, "Dimension of BK matrix must be at least 1"
    # Base case
    if n == 1:
        return np.ones((1, 1), dtype=np.int8)

    # Recursive definition
    smaller_bk_matrix = _bk_matrix_power2(n - 1)
    top_right = np.zeros((2 ** (dims - 2), 2 ** (dims - 2)), dtype=np.int8)
    top_half = np.concatenate((smaller_bk_matrix, top_right), axis=1)

    bottom_left = np.concatenate(
        (np.zeros(((2 ** (dims - 2)) - 1, 2 ** (dims - 2)), dtype=np.int8), 
        np.ones((1, 2 ** (dims - 2)), dtype=np.int8)), 
        axis=0
    )
    bottom_half = np.concatenate((bottom_left, smaller_bk_matrix), axis=1)
    
    # Combine top and bottom half
    return np.concatenate((top_half, bottom_half), axis=0)
```

renatomello replied:
If `dims` must be a power of :math:`2`, there could be an `assert` verifying that.

chmwzc committed:
Update src/qibochem/ansatz/hf_reference.py

Co-authored-by: Renato M S Farias <renato.msf@gmail.com>

chmwzc committed:
Update src/qibochem/ansatz/hf_reference.py

Co-authored-by: Renato M S Farias <renato.msf@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

chmwzc committed:
Fix tests

Also rename some instances of n_qubits -> nqubits and n_electrons -> nelectrons (in-progress)

chmwzc committed:
Merge branch 'main' into encodings

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

chmwzc committed:
Merge branch 'main' into encodings

chmwzc committed:
Add type hinting

chmwzc committed:
Minor edits to HF circuit

chmwzc committed:
Refactor Givens excitation circuit

Now uses GIVENS and GeneralizedRBS gates directly

chmwzc committed:
Rename util.py to utils.py

chmwzc committed:
Forgot to update `util` in the modules

chmwzc committed:
Initial draft of combined ansatz module

chmwzc committed:
Add hardware efficient ansatz

chmwzc committed:
Move ucc_circuit code and tests

chmwzc committed:
Add qeb_circuit and test module

chmwzc committed:
Add/Update Givens circuit ansatz and tests

chmwzc committed:
Merge branch 'main' into encodings

chmwzc committed:
Remove old test files

chmwzc committed:
Remove old ansatz files

chmwzc committed:
Minor edits/fixes after removing old files

chmwzc committed:
Minor bugfixes

chmwzc committed:
Minor fixes for code coverage

chmwzc committed:
Migrate basis_rotation to ansatz.py (in-progress)

chmwzc committed:
Continue migrating basis rotation ansatz code

With some refactoring done too

chmwzc committed:
Refactor basis rotation code

chmwzc committed:
Finish basis rotation ansatz and tests

chmwzc committed:
Remove old module and tests

chmwzc committed:
Move symmetry preserving ansatz and tests

chmwzc committed:
Cleanup symmetry-ansatz stuff

chmwzc committed:
Update ansatz utils

Minor edits to type hints too

chmwzc committed:
Wrap up individual ansatz code

Started work on docs

chmwzc committed:
Tidy up docs and sketch circuit_ansatz function

chmwzc committed:
Finish circuit_ansatz function

chmwzc committed:
Add test for circuit_ansatz

chmwzc committed:
Fix code coverage

chmwzc committed:
Remove include_hf from basis_rotation_circuit

To be consistent with ucc/qeb/givens

chmwzc committed:
Update docs and fix tests (hopefully)

chmwzc committed:
Add Hamming-weight encoder into ansatzes

chmwzc committed:
Add Hamming-weight ansatz to __init__.py

chmwzc commented:
Yay, finally finished this. The main changes are:

## Refactoring:
- Migrated all existing circuit ansatzes into `ansatz.py`, helper functions to `_ansatz.py`, utility functions in `utils.py`
- Replaced hardcoded decompositions in the `givens_circuit` function (previous name `givens_excitation_circuit`) with `gates.GIVENS` and `gates.GeneralizedRBS`
  - Sidenote: As far as I can tell, UCC can't be expressed using those gates.

## Added:
  - Convenience function: `circuit_ansatz()` that takes in a `Molecule` object and builds a circuit ansatz.
  - Function to return a `Circuit` for the basis rotation ansatz; previously, was just returning a list of gates
  - Hamming-weight preserving ansatz: The `hamming_weight_encoder` from Qibo

renatomello commented:
> Yay, finally finished this. The main changes are:
> ## Refactoring:
> 
>     * Migrated all existing circuit ansatzes into `ansatz.py`, helper functions to `_ansatz.py`, utility functions in `utils.py`
> 
>     * Replaced hardcoded decompositions in the `givens_circuit` function (previous name `givens_excitation_circuit`) with `gates.GIVENS` and `gates.GeneralizedRBS`
>       
>       * Sidenote: As far as I can tell, UCC can't be expressed using those gates.
> 
> 
> ## Added:
> 
>     * Convenience function: `circuit_ansatz()` that takes in a `Molecule` object and builds a circuit ansatz.
> 
>     * Function to return a `Circuit` for the basis rotation ansatz; previously, was just returning a list of gates
> 
>     * Hamming-weight preserving ansatz: The `hamming_weight_encoder` from Qibo

@chmwzc Not that it is going to make a whole lot of difference, but the last `CNOT - RZ - CNOT` block in the UCC decomposition is an `qibo.gates.RZZ` gate with the same angle.

chmwzc commented:
@renatomello Do you still want to review this, or can I merge it in as it is?

renatomello commented on src/qibochem/ansatz/_ansatz.py:
```diff
@@ -0,0 +1,294 @@
+    return min_bk_matrix[:dims, :dims]
+
+
+def _expi_pauli(n_qubits: int, pauli_string: str, theta: float, **kwargs) -> Circuit:
```
Minor: everywhere else, including this file, uses `nqubits` instead of `n_qubits`

renatomello reviewed (approved):
I have comments on more structural / design issues, but they are out of the scope of this PR.

chmwzc committed:
Rename `n_qubits` to `nqubits`
